agent/proxycfg/upstreams.go

```diff
@@ -92,55 +90,60 @@ func (s *handlerUpstreams) handleUpdateUpstreams(ctx context.Context, u cache.Up
 		}
 		upstreamsSnapshot.WatchedUpstreamEndpoints[uid][targetID] = resp.Nodes
 
-		var passthroughAddrs map[string]ServicePassthroughAddrs
+		if s.kind != structs.ServiceKindConnectProxy || s.proxyCfg.Mode != structs.ProxyModeTransparent {
+			return nil
+		}
 
-		for _, node := range resp.Nodes {
-			if snap.Proxy.Mode == structs.ProxyModeTransparent && node.Service.Proxy.TransparentProxy.DialedDirectly {
-				if passthroughAddrs == nil {
-					passthroughAddrs = make(map[string]ServicePassthroughAddrs)
+		if _, ok := upstreamsSnapshot.PassthroughUpstreams[uid]; ok {
+			for addr := range upstreamsSnapshot.PassthroughUpstreams[uid][targetID] {
+				if indexed := upstreamsSnapshot.PassthroughIndices[addr]; indexed.targetID == targetID && indexed.upstreamID.Matches(uid) {
+					delete(upstreamsSnapshot.PassthroughIndices, addr)
 				}
+			}
+			upstreamsSnapshot.PassthroughUpstreams[uid][targetID] = make(map[string]struct{})
+		}
 
-				svc := node.Service.CompoundServiceName()
-
-				// Overwrite the name if it's a connect proxy (as opposed to Connect native).
-				// We don't reference the proxy name directly for things like SNI, but rather the name
-				// of the destination. The enterprise meta of a proxy will always be the same as that of
-				// the destination service, so that remains intact.
-				if node.Service.Kind == structs.ServiceKindConnectProxy {
-					dst := node.Service.Proxy.DestinationServiceName
-					if dst == "" {
-						dst = node.Service.Proxy.DestinationServiceID
-					}
-					svc.Name = dst
-				}
+		passthroughs := make(map[string]struct{})
+
+		for _, node := range resp.Nodes {
+			if !node.Service.Proxy.TransparentProxy.DialedDirectly {
+				continue
+			}
 
-				sni := connect.ServiceSNI(svc.Name, "", svc.NamespaceOrDefault(), svc.PartitionOrDefault(), snap.Datacenter, snap.Roots.TrustDomain)
+			// Make sure to use an external address when crossing partitions.
+			// Datacenter is not considered because transparent proxies cannot dial other datacenters.
+			isRemote := !structs.EqualPartitions(node.Node.PartitionOrDefault(), s.proxyID.PartitionOrDefault())
+			addr, _ := node.BestAddress(isRemote)
 
-				spiffeID := connect.SpiffeIDService{
-					Host:       snap.Roots.TrustDomain,
-					Partition:  svc.PartitionOrDefault(),
-					Namespace:  svc.NamespaceOrDefault(),
-					Datacenter: snap.Datacenter,
-					Service:    svc.Name,
-				}
+			existing := upstreamsSnapshot.PassthroughIndices[addr]
 
-				svcUID := NewUpstreamIDFromServiceName(svc)
-				if _, ok := upstreamsSnapshot.PassthroughUpstreams[svcUID]; !ok {
-					upstreamsSnapshot.PassthroughUpstreams[svcUID] = ServicePassthroughAddrs{
-						SNI:      sni,
-						SpiffeID: spiffeID,
+			// The modify index is chosen based on the logic from BestAddress
+			csnIdx := node.Service.ModifyIndex
+			if node.Service.Address == "" {
+				csnIdx = node.Node.ModifyIndex
+			}
```

Comment:
You could do two things to make this easier to track:

1. move this into a sibling method `BestAddressModifyIndex`
2. change the return value of `BestAddress` to be `(uint64 /*raft index*/, string /*addr*/, int /*port*/)`